### PR TITLE
Fix for mention in link followed by punctuation

### LINKED ISSUE #91: Punctuation following an email address in Markdown fields causes it not to be parsed as email
(This came from  HabitRPG/habitica#12437)
The following test case fails
```javascript
    it('does not parse mentions in e-mail addresses ending in punctuation', () => {
      const text = 'send an e-mail to nevermind@user.com!';

      const mentionTokens = findMentionTokens(md.parse(text));

      expect(mentionTokens.length).to.equal(0);
    });
```

### LINKED ISSUE #12437: Punctuation following an email address in Markdown fields causes it not to be parsed as email
[//]: # (Before logging this issue, please post to the Report a Bug guild from the Habitica website's Help menu. Most bugs can be handled quickly there. If a GitHub issue is needed, you will be advised of that by a moderator or staff member -- a player with a dark blue or purple name. It is recommended that you don't create a new issue unless advised to.)

[//]: # (Bugs in the mobile apps can also be reported there.)

[//]: # (If you have a feature request, use "Help > Request a Feature", not GitHub or the Report a Bug guild.)

[//]: # (For more guidelines see https://github.com/HabitRPG/habitica/issues/2760)

[//]: # (Fill out relevant information - UUID is found from the Habitia website at User Icon > Settings > API)
### Description
[//]: # (Describe bug in detail here. Include screenshots if helpful.)

Ordinarily, an email address typed in chat, private messages, etc. is parsed and rendered as a `mailto:` link. However, if you follow it with punctuation, like a period or a hyphen used to make a dash, the email is not understood as such and gets a partial highlight meant to call out usernames:

image

#### Console Errors
[//]: # (Include any JavaScript console errors here.)

None.

### PR DESCRIPTION
**Pull request by mbroshi**
Added a test for an edge case that was not parsing properly. Changed isLinkOrEmail to check if the link contains a mention. All tests (including new one) are passing.

Fixes https://github.com/HabitRPG/habitica/issues/12437
Fixes https://github.com/HabitRPG/habitica-markdown/issues/91

### CODE REVIEW

File: lib/mentionsPlugin.js [hunk 1]
```diff
@@ -30,13 +30,19 @@ function parseMentions (state, silent) {
     return false;
   }
 
-  const text = state.src.slice(state.pos);
-  const match = text.match(mentionRegex);
-
-  if (!match || partOfLinkOrEmail(state.src.substr(0, state.pos), text, state.md)) {
+  const textUntilAt = state.src.slice(0, state.pos);
+  const textAfterAt = state.src.slice(state.pos);
+
```
Comment (paglias) [inline lib/mentionsPlugin.js]: eslint is showing an extra trailing space at this line 
Schermata 2020-08-25 alle 17 23 50

File: test/lib/mentionsPlugin.test.js [hunk 2]
```diff
@@ -39,6 +39,14 @@ describe('mentionsPlugin', () => {
       expect(mentionTokens.length).to.equal(0);
     });
 
+    it('does not parse mentions in e-mail addresses ending in punctuation', () => {
```
Comment (paglias) [inline test/lib/mentionsPlugin.test.js:42]: @mbroshi this test is correctly passing but I tried with the example posted in the original issue (`admin@habitica.com-test)` and it's still being counted as a mention

**Review by paglias — CHANGES_REQUESTED** [review]
left a couple of comments!

File: test/lib/mentionsPlugin.test.js (hunk 2 above)
Comment (mbroshi) [reply test/lib/mentionsPlugin.test.js:42]: Thanks for this catch! You are totally correct. I just opened an issue on linkify-it: https://github.com/markdown-it/linkify-it/issues/93

Other punctuation seems to be fine, but not hyphens. I will wait to hear back from them.

**Comment by mbroshi** [comment]
@paglias Thank you kindly for the feedback! I have a couple questions for you:
1. How can I test my changes out locally in Habitica? I tried editing habitica/package.json and replaced `"habitica-markdown":"2.02"` with `"habitica-markdown":"file:../habitica-markdown"` which seems to have correctly grabbed my local changes into habitica/node_modules, but when I look in my browser, I still see the old habitica-markdown in the chunk-vendors.js file.
2. Should I wait on the linkify issue to resolve? I have have workaround, which is to strip trailing hyphens in `isLinkOrEmail` before checking for a link. That appears to be the only punctuation that trips up linkify. It's harmless, but slightly hacky. The email link will not render with a mailto link in the browser, but at least it no longer has the spurious mention in it. I could submit that while we wait for the linkify issue.

**Comment by mbroshi** [comment]
One final question: I see `md.isLinkOrEmail = ...` in `createMdInstance.js` line 27. Is that used anywhere? I can delete it and all tests still pass. I'm trying to see if I need to update that line as well. Thanks!

**Comment by paglias** [comment]
Looking at the list of valid TLDs here https://www.iana.org/domains/root/db there are none with punctuation so I don't think we can expect `linkify` to support them and it's an extra rare use case so I think your current fix that supports for example exclamation points is good enough

**Comment by mbroshi** [comment]
Based on your comment, I am leaving the rest of my changes as-is. Could you please clarify these two additional questions:
1. Do I need to update line 27 of `createMdInstance.js`? I don't see it used anywhere else.
2. Should I also update the readme? It mentions isLinkOrEmail, but I've now made a breaking x-change for that function that requires a mention to be passed in.

**Comment by paglias** [comment]
> How can I test my changes out locally in Habitica? I tried editing habitica/package.json and replaced "habitica-markdown":"2.02" with "habitica-markdown":"file:../habitica-markdown" which seems to have correctly grabbed my local changes into habitica/node_modules, but when I look in my browser, I still see the old habitica-markdown in the chunk-vendors.js file.

Have you solved this issue?

> Do I need to update line 27 of createMdInstance.js? I don't see it used anywhere else.

That should stay, it's not used internally in the library but added to the `md` object so that it can be used in other modules consuming this one

**Comment by paglias** [comment]
Regarding the breaking change, I don't think it's necessary to mention it in the README as `isLinkOrEmail` wasn't documented before either, but I'll probably have to release version 3 of the module when this is merged

File: lib/mentionsPlugin.js [hunk 3]
```diff
@@ -30,13 +30,19 @@ function parseMentions (state, silent) {
...
+  const textAfterAt = state.src.slice(state.pos);
 
-  if (!match || partOfLinkOrEmail(state.src.substr(0, state.pos), text, state.md)) {
+  const match = textAfterAt.match(mentionRegex);
+  if (!match) {
     return false;
   }
+
   const mention = match[0];
+  if (partOfLinkOrEmail(mention, textUntilAt, textAfterAt, state.md)) {
```
Comment (benkelaar) [inline lib/mentionsPlugin.js]: What's the benefit of splitting up this conditional?

File: lib/util/isLinkOrEmail.js [hunk 4]
```diff
@@ -1,10 +1,10 @@
 /*
  * Uses `linkify` to test whether the @mention is part of a link or e-mail address
  */
-function isLinkOrEmail (text, linkify) {
+function isLinkOrEmail (mention, text, linkify) {
```
Comment (benkelaar) [inline lib/util/isLinkOrEmail.js]: Perhaps rename to `isMentionInLinkOrEmail`

File: lib/mentionsPlugin.js [hunk 5]
```diff
@@ -7,7 +7,7 @@ const isLinkOrEmail = require('./util/isLinkOrEmail');
 const mentionRegex = /^@[\w-]+/;
 const whiteSpaceRegex = /\s+/;
 
-function partOfLinkOrEmail (pretext, text, md) {
+function partOfLinkOrEmail (mention, pretext, text, md) {
```
Comment (benkelaar) [inline lib/mentionsPlugin.js]: Perhaps use the same parameter names as the variable names in the calling context to improve continuity.

**Review by benkelaar — CHANGES_REQUESTED** [review]
Great improvement, thanks! Left a couple of small cosmetic comments.

**Comment by benkelaar** [comment]
> One final question: I see `md.isLinkOrEmail = ...` in `createMdInstance.js` line 27. Is that used anywhere? I can delete it and all tests still pass. I'm trying to see if I need to update that line as well. Thanks!

Ah good point, is no longer used in habitica, it now uses the parsing of this module directly.  So that method can indeed be removed from the exports and doesn't need to be its own module as well, can probably be inlined as a private method in `mentionsPlugin.js`.

@paglias But yeah, you're right that would make this version 3.

**Comment by benkelaar** [comment]
> > How can I test my changes out locally in Habitica? I tried editing habitica/package.json and replaced "habitica-markdown":"2.02" with "habitica-markdown":"file:../habitica-markdown" which seems to have correctly grabbed my local changes into habitica/node_modules, but when I look in my browser, I still see the old habitica-markdown in the chunk-vendors.js file.
> 
> Have you solved this issue?

I mentioned how to do this in the Aspiring Coders guild, see also: https://medium.com/@AidThompsin/how-to-npm-link-to-a-local-version-of-your-dependency-84e82126667a#:~:text=The%20TL%3BDR%3A%20clone%20the,now%20using%20the%20local%20version.

I mentioned there that it would be good to add to the `README.md`, but now that I think about it would probably be better in a `CONTRIBUTING.md`. However, not sure if you want to pick that up, could be temporarily added to the README.

File: lib/mentionsPlugin.js (hunk 3 above)
Comment (mbroshi) [reply lib/mentionsPlugin.js]: Thanks @benkelaar for the careful feedback!

I now need `match[0]` in the call to `partOfLinkOrEmail`. The alternative is `if (!match && partOfLinkOrEmail(match[0], ...)` but in that case it's not clear what `match[0]` is without the variable name.

File: lib/util/isLinkOrEmail.js (hunk 4 above)
Comment (mbroshi) [reply lib/util/isLinkOrEmail.js]: Definitely! I like the suggestion.

File: lib/mentionsPlugin.js (hunk 5 above)
Comment (mbroshi) [reply lib/mentionsPlugin.js]: Make sense--will do.

File: lib/mentionsPlugin.js (hunk 3 above)
Comment (benkelaar) [reply lib/mentionsPlugin.js]: Ah yeah, so it's a consideration between duplication of `match[0]` and duplication of `return false`.

I personally find that limiting the amount of separate possible exit-points of a function makes it easier to reason about program flow, which I think is worth the duplication of `match[0]` (especially since `if (!match || queryUsing(match.parameter))` is a fairly common pattern for null-checking in javascript).

The variable name is of course present as parameter name in `partOfLinkOrEmail`, which is where it is relevant.  However in the end it's a flavour call.

You could possibly turn the call into a call with [faux-named-parameters](https://exploringjs.com/impatient-js/ch_callables.html#named-parameters), but I don't think that'll increase the readability.  Might be a consideration though.

Comment (benkelaar) [reply lib/mentionsPlugin.js]: Hmm, now that I think about it though, that would turn it in to:

```javascript
function partOfLinkOrEmail ({ mention, textUntilAt, textAfterAt, md }) {
// ...

// and then later
if (!match || partOfLinkOrEmail({mention: match[0], textUntilAt, textAfterAt, md: state.md}))
```

Which might just be exactly what you need. Haven't tried it yet though, perhaps you don't even need to provide the name for `state.md` either, but not sure.

Edit: Might trip up the linter though, if that's the case it's probably not worth it.

Comment (mbroshi) [reply lib/mentionsPlugin.js]: The linter was not fond of that named parameter approach without line breaks before/after the curly braces, which looks hideous. I'll switch to the `match[0]` approach since that is more idiomatic. FWIW, I have been mostly coding in Go lately, where early exit on edge cases (leaving only the happy path) is considered idiomatic and more readable.

**Comment by mbroshi** [comment]
Thanks @benkelaar for the tip on getting habitica-markdown integrated with habitica! I got it all linked locally now, and I was able to qualify in my local Habitica chat that email addresses ending in punctuation render correctly. Except followed by hyphens, that is. That issue is still open in linkify-it. From our end, since that is the only punctuation that does not render correctly, the workaround would be to strip trailing hyphens in the `isMentionInLinkOrEmail` function.

OK, so this is actually my very first pull request in Github. Is there further action required on my part? Thanks @benkelaar and @paglias for the careful reviews and walking me through contributing!

**Comment by paglias** [comment]
thanks @mbroshi and thanks @benkelaar for helping around, I'm going to merge the PR and release version 3 of the module, it will go live next week after we do our weekly dependencies upgrades

@mbroshi do you have an Habitica username? If so post it here so I can award a contribution tier for the PR

**Comment by mbroshi** [comment]
> @mbroshi do you have an Habitica username? If so post it here so I can award a contribution tier for the PR
Thanks @paglias! My Habitica username is @wslindon

**Comment by paglias** [comment]
@mbroshi I'm awarding your first contributor tier!

Note that further tiers require increasing amounts of work from one to the next, so Tier 2 may require a couple of PRs or a larger PR to attain. But keep helping out and we'll express our gratitude accordingly!
